src/lib/foreign-field.ts

```diff
@@ -0,0 +1,496 @@
+      return new ForeignField(z);
+    }
+
+    /**
+     * Finite field multiplication
+     *
+     * ```ts
+     * x.mul(y); // x*y mod p
+     * ```
+     */
+    mul(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        let z = mod(this.toBigInt() * toFp(y), p);
+        return new ForeignField(z);
+      }
+      let z = Snarky.foreignField.mul(this.value, toVar(y), pMl);
+      return new ForeignField(z);
+    }
+
+    /**
+     * Multiplicative inverse in the finite field
+     *
+     * ```ts
+     * let z = x.inv(); // 1/x mod p
+     * z.mul(x).assertEquals(1);
+     * ```
+     */
+    inv(): ForeignField {
+      if (this.isConstant()) {
+        let z = inverse(this.toBigInt(), p);
+        if (z === undefined) {
+          if (this.toBigInt() === 0n) {
+            throw Error('ForeignField.inv(): division by zero');
+          } else {
+            // in case this is used with non-prime moduli
+            throw Error('ForeignField.inv(): inverse does not exist');
+          }
+        }
+        return new ForeignField(z);
+      }
+      let z = Provable.witness(ForeignField, () => this.toConstant().inv());
+
+      // in unsafe mode, `witness` didn't constrain z to be a valid field element
+      if (unsafe) z.assertValidElement();
+
+      // check that x * z === 1
+      // TODO: range checks added by `mul` on `one` are unnecessary, since we already assert that `one` equals 1
+      let one = Snarky.foreignField.mul(this.value, z.value, pMl);
+      new ForeignField(one).assertEquals(new ForeignField(1));
+
+      return z;
+    }
+
+    // convenience methods
+
+    /**
+     * Assert equality with a ForeignField-like value
+     *
+     * ```ts
+     * x.assertEquals(0, "x is zero");
+     * ```
+     */
+    assertEquals(y: ForeignField | bigint | number, message?: string) {
+      try {
+        if (this.isConstant() && isConstant(y)) {
+          let x = this.toBigInt();
+          let y0 = toFp(y);
+          if (x !== y0) {
+            throw Error(`ForeignField.assertEquals(): ${x} != ${y0}`);
+          }
+        }
+        return Provable.assertEqual(ForeignField, this, ForeignField.from(y));
+      } catch (err) {
+        throw withMessage(err, message);
+      }
+    }
+
+    /**
+     * Check equality with a ForeignField-like value
+     *
+     * ```ts
+     * let isXZero = x.equals(0);
+     * ```
+     */
+    equals(y: ForeignField | bigint | number) {
+      if (this.isConstant() && isConstant(y)) {
+        return new Bool(this.toBigInt() === toFp(y));
+      }
+      return Provable.equal(ForeignField, this, ForeignField.from(y));
+    }
+
+    // bit packing
+
+    /**
+     * Unpack a field element to its bits, as a `Bool[]` array.
+     *
+     * This method is provable!
+     */
+    toBits(length = sizeInBits) {
+      checkBitLength('ForeignField.toBits()', length, sizeInBits);
+      let [l0, l1, l2] = this.toFields();
```

Comment:
Just for clarity, `l0` is the least significant limb?

Reply:
yes!